test/modality.spec.ts

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+import { suite, test } from 'mocha';
+import { expect } from 'chai';
+import {
+    InputModality,
+    ModalityEvaluationDefaults,
+    OutputModality,
+    ResponseStyle,
+} from '../src/modality/ModalityEvaluation';
+import { HandlerInput } from 'ask-sdk-core';
+import { TestInput } from '../src/utils/testSupport/TestingUtils';
+import { ControlHandler } from '../src/runtime/ControlHandler';
+import { SkillInvoker } from '../src/utils/testSupport/SkillInvoker';
+import { ControlManager, ControlManagerProps } from '../src/controls/ControlManager';
+import { ControlInput } from '../src/controls/ControlInput';
+import { ContainerControl } from '../src/controls/ContainerControl';
+import { ControlResultBuilder } from '../src/controls/ControlResult';
+
+suite('== Modality evaluation and tracking scenarios ==', () => {
+    const touchInput = TestInput.simpleUserEvent([]);
+    const voiceInput = TestInput.of('VoiceIntent');
+    let manager: TestControlManager;
+    let handler: ControlHandler;
+    let rootControl: TestControl;
+    let lastHistory: InputModality[] | undefined;
+    let invoker: SkillInvoker;
+
+    beforeEach(() => {
+        manager = new TestControlManager({
+            responseStyleEvaluator,
+        });
+        handler = new ControlHandler(manager);
+        rootControl = manager.createControlTree();
+        rootControl.lastKnownModality = undefined;
+        lastHistory = undefined;
+        invoker = new SkillInvoker(handler);
+    });
+
+    test('Input and output modality are properly evaluated', async () => {
+        await invoker.invoke(touchInput);
+        expect(await rootControl.lastKnownModality).to.equal(OutputModality.SCREEN);
+
+        await invoker.invoke(voiceInput);
+        expect(await rootControl.lastKnownModality).to.equal(OutputModality.VOICE);
+    });
+
+    test('Input modality history is properly tracked', async () => {
+        await invoker.invoke(touchInput);
+        await invoker.invoke(voiceInput);
+        expect(lastHistory).to.deep.equal(['touch', 'voice']);
+    });
+
+    function responseStyleEvaluator(input: HandlerInput, history: InputModality[]): ResponseStyle {
+        lastHistory = JSON.parse(JSON.stringify(history));
```

Comment:
Why is `lastHistory` evaluated and un-used here?

Reply:
This is here to expose some of the internals of the evaluator so that assertions can be made by the unit tests; it's used on line 62. I could alternatively set some data we want asserted in the test and have the assertion happen here, if you think that will be clearer. Open to other ideas as well.

Reply:
No issues I missed the line 62 usage.
Thanks for the explanation